### PR TITLE
Remove StoreAdapter

### PR DESCRIPTION
**Pull request by ahmetmircik**
__Goal:__
To introduce new HDRecord types, i'm trying to remove HDRecord usages from other packages, like query related class package. This is to increase isolation and making further changes easy.

__Main Modifications:__
- Removed `StoreAdapter` and relevant references to it.
- Filtered out expired entries in Indexes.query method.

ee counterpart: https://github.com/hazelcast/hazelcast-enterprise/pull/3860

### CODE REVIEW

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/2975/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/2975/Hazelcast-pr-builder-2975.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Errors: 
[ERROR]   IndexConcurrencyTest.testIndexCreationAndQueryDeterministicConcurrency:105->assertQuery:128->testIndexCreationAndQueryDeterministicConcurrency:105->assertQuery:128 ? OperationTimeout
[INFO] 
[ERROR] Tests run: 35381, Failures: 0, Errors: 1, Skipped: 979
[INFO] 

[ERROR] There are test failures.
</details>
</pre>

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/2979/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/2979/Hazelcast-pr-builder-2979.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.1:checkstyle (default) on project hazelcast: An error has occurred in Checkstyle report generation. Failed during checkstyle execution: There is 1 error reported by Checkstyle 8.29 with /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/checkstyle/checkstyle.xml ruleset. -> [Help 1]
--------------------------
---------ERRORS-----------
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/util/IterableUtil.java:73:23: Variable 'next' explicitly initialized to 'null' (default value for its type). [ExplicitInitialization]
--------------------------
</details>
</pre>

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/2984/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/2984/Hazelcast-pr-builder-2984.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] Failed to execute goal com.github.spotbugs:spotbugs-maven-plugin:3.1.12:check (default) on project hazelcast: failed with 1 bugs and 0 errors -> [Help 1]
--------------------------
</details>
</pre>

File: hazelcast/src/main/java/com/hazelcast/internal/util/IterableUtil.java [hunk 1]
```diff
@@ -61,6 +69,63 @@ public void remove() {
...
+            private T next;
+
+            @Override
+            public boolean hasNext() {
+                boolean hasNext = false;
+                while (givenIterator.hasNext()) {
+                    T temp = givenIterator.next();
+                    if (filter.test(temp)) {
+                        next = temp;
+                        hasNext = true;
```
Comment (petrpleshachkov) [inline hazelcast/src/main/java/com/hazelcast/internal/util/IterableUtil.java:91]: Do we really need `hasNext` local variable here? `return true` seems enough.

File: hazelcast/src/main/java/com/hazelcast/internal/util/IterableUtil.java [hunk 2]
```diff
@@ -61,6 +69,63 @@ public void remove() {
...
+                        break;
+                    }
+                }
+                return hasNext;
+            }
+
+            @Override
+            public T next() {
+                if (next == null) {
```
Comment (petrpleshachkov) [inline hazelcast/src/main/java/com/hazelcast/internal/util/IterableUtil.java:100]: We should `throw NoSuchElementException` only if there is no next element not expecting that the user called `hasNext()` before that. I'd put it like
if (!hasNext()) {
    throw new NoSuchElementException()
}

File: hazelcast/src/main/java/com/hazelcast/internal/util/IterableUtil.java [hunk 3]
```diff
@@ -61,6 +69,63 @@ public void remove() {
...
+    public static <T> Iterable<T> filter(Iterable<T> iterable, Predicate<T> filter) {
+        Iterator<T> givenIterator = iterable.iterator();
+
+        Iterator<T> filteringIterator = new Iterator<T>() {
+            private T next;
+
+            @Override
+            public boolean hasNext() {
+                boolean hasNext = false;
+                while (givenIterator.hasNext()) {
```
Comment (petrpleshachkov) [inline hazelcast/src/main/java/com/hazelcast/internal/util/IterableUtil.java:87]: If I call `hasNext()` multiple times it shouldn't advance the iterator, but it seems it does.

File: hazelcast/src/main/java/com/hazelcast/map/impl/MapContainer.java [hunk 4]
```diff
@@ -158,7 +159,24 @@ public Indexes createIndexes(boolean global) {
...
                 .partitionCount(partitionCount)
-                .build();
+                .resultFilter(queryableEntry -> hasNotExpired(queryableEntry)).build();
+    }
+
+    /**
+     * @return {@code true} if queryableEntry has
+     * not expired, otherwise returns {@code false}
+     */
+    private boolean hasNotExpired(QueryableEntry queryableEntry) {
```
Comment (petrpleshachkov) [inline hazelcast/src/main/java/com/hazelcast/map/impl/MapContainer.java:169]: Is this method called on the partition thread?

Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/map/impl/MapContainer.java:170]: yes, when it is not global index or i can't find counter example at least. I have added check `assertRunningOnPartitionThread` when index is not global.

File: hazelcast/src/main/java/com/hazelcast/internal/util/IterableUtil.java (hunk 3 above)
Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/internal/util/IterableUtil.java:93]: good catch, fixed.

File: hazelcast/src/main/java/com/hazelcast/internal/util/IterableUtil.java (hunk 2 above)
Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/internal/util/IterableUtil.java]: fixed.

File: hazelcast/src/main/java/com/hazelcast/internal/util/IterableUtil.java (hunk 1 above)
Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/internal/util/IterableUtil.java]: no need, fixed.

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/2994/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/2994/Hazelcast-pr-builder-2994.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.1:checkstyle (default) on project hazelcast: An error has occurred in Checkstyle report generation. Failed during checkstyle execution: There are 2 errors reported by Checkstyle 8.29 with /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/checkstyle/checkstyle.xml ruleset. -> [Help 1]
--------------------------
---------ERRORS-----------
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast/src/main/java/com/hazelcast/internal/util/IterableUtil.java:84:25: Variable 'hasNextCallCount' explicitly initialized to '0' (default value for its type). [ExplicitInitialization]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast/src/main/java/com/hazelcast/internal/util/IterableUtil.java:85:25: Variable 'nextCallCount' explicitly initialized to '0' (default value for its type). [ExplicitInitialization]
--------------------------
</details>
</pre>

File: hazelcast/src/main/java/com/hazelcast/internal/util/IterableUtil.java [hunk 5]
```diff
@@ -61,6 +69,72 @@ public void remove() {
...
+     * @return a new iterable object which
+     * has an iterator capable of filtering elements
+     */
+    public static <T> Iterable<T> filter(Iterable<T> iterable, Predicate<T> filter) {
+        Iterator<T> givenIterator = iterable.iterator();
+
+        @SuppressWarnings("checkstyle:anoninnerlength")
+        Iterator<T> filteringIterator = new Iterator<T>() {
+            private T next;
+            private int hasNextCallCount;
```
Comment (petrpleshachkov) [inline hazelcast/src/main/java/com/hazelcast/internal/util/IterableUtil.java:84]: Not sure that we need these 2 variables. Why not using just `next`? If it is not `null` the hasNext() returns `true` immediately. The `next()` will first calls `hasNext()` and if it exists, returns it (assigning `next` to `null`). Something similar to the `HDBplusTree#EntryIterator`.

Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/internal/util/IterableUtil.java:84]: but in that case consecutive `hasNext` calls advance state, i don't want it to happen.

Comment (petrpleshachkov) [reply hazelcast/src/main/java/com/hazelcast/internal/util/IterableUtil.java:84]: No, it'll check that `next` variable is not null and return quickly.

Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/internal/util/IterableUtil.java]: Oops i see now, my test was wrong, also fixed it, thx.

**Review by petrpleshachkov — APPROVED** [review]

File: hazelcast/src/main/java/com/hazelcast/query/impl/BaseIndexStore.java [hunk 6]
```diff
@@ -145,13 +145,6 @@ private Comparable sanitizeScalar(Object input) {
         }
     }
 
-    void markIndexStoreExpirableIfNecessary(QueryableEntry record) {
```
Comment (mmedenjak) [inline hazelcast/src/main/java/com/hazelcast/query/impl/BaseIndexStore.java:148]: I'm a bit lost here. Now we don't update the `isIndexStoreExpirable` field anywhere but we still use the return value from `BaseIndexStore#isExpirable` in three places. Can you elaborate on this change?

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/3054/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/3054/Hazelcast-pr-builder-3054.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] Failed to execute goal com.github.spotbugs:spotbugs-maven-plugin:3.1.12:check (default) on project hazelcast: failed with 2 bugs and 0 errors -> [Help 1]
--------------------------
</details>
</pre>

**Review by mmedenjak — COMMENTED** [review]
Went through the whole PR, looks good except for that one comment.

File: hazelcast/src/main/java/com/hazelcast/query/impl/BaseIndexStore.java (hunk 6 above)
Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/query/impl/BaseIndexStore.java:148]: leftover, removed that method and related stuff.

Comment (mmedenjak) [reply hazelcast/src/main/java/com/hazelcast/query/impl/BaseIndexStore.java:148]: Still, can you elaborate a bit on how it's safe to remove all of that code? With the removal, the dependant code now acts as if `isExpirable() == false`, always.

Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/query/impl/BaseIndexStore.java:148]: All filtering for expiry is done in one place now. In `Indexes#query` method,`resultFilter` will check for expiry. See the related part in that method:
```
   if (result != null && resultFilter != null) {
            return IterableUtil.filter(result, resultFilter);
        } else {
            return result;
        }
```

**Review by mmedenjak — APPROVED** [review]
Perfect. And great that you managed to remove this much complexity.

File: hazelcast/src/main/java/com/hazelcast/query/impl/BaseIndexStore.java (hunk 6 above)
Comment (mmedenjak) [reply hazelcast/src/main/java/com/hazelcast/query/impl/BaseIndexStore.java:148]: Thanks for the explanation!

**Pull request merged**